### PR TITLE
Added script for lambda code archive building + adapted main template for deployment

### PR DESCRIPTION
**Pull request by BarcoMasile**
## Description
This PR introduces a script to package PCUI code as a zip archive for Lambda deployment.
It also edits the main CloudFormation template `parallelcluster-ui.yaml` to allow the lambda deployment and correctly set everything up (otherwise it would have not been possible to test the changes).
This assumes that the Lambda code archive is available in the same bucket where infrastructure CloudFormation templates are available.

Right now the script outputs all the info from the various commands it executes.
The produced archive will be available at the path from the two specified mandatory parameters.
Check the `build_lambda_archive.sh` top file comments for example on its usage.

**NB**: Right now only tests files are removed from the package. In a future PR we will separate dev dependencies and prod dependencies, so that Boto3 library and pytest (and other dev dependencies) will not be included in the bundle, making it lighter.

The main template has been stripped of several resources that are not needed anymore:
- *PrivateEcrRepository*
- *ImageBuilderInstanceRole*
- *ImageBuilderInstanceProfile*
- *InfrastructureConfiguration*
- *InfrastructureConfigurationSecurityGroup*
- *EcrImageDeletionLambda*
- *EcrImageDeletionLambdaLogGroup*
- *EcrImageDeletionLambdaRole*
- *EcrImagesRemover*
- *PrivateEcrRepository*
- *EcrImagePipeline*
- *EcrImageRecipe*
- *EcrImage*

## How Has This Been Tested?
This has been tested manually on my personal account, by generating the zip archive with the different options and by deploying the created archive using the edited CloudFormation template.

In order to increase the likelihood of your contribution being accepted, please make sure you have read both the [Contributing Guidelines](../CONTRIBUTING.md) and the [Project Guidelines](../PROJECT_GUIDELINES.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: infrastructure/parallelcluster-ui.yaml [hunk 1]
```diff
@@ -3,10 +3,10 @@ Parameters:
     Description: Email address of administrative user to setup by default (only with new Cognito instances).
     Type: String
     Default: ''
-  PublicEcrImageUri:
-    Description: When specified, the URI of the Docker image for the Lambda of the ParallelCluster UI container
+  PCUIVersion:
```
Comment (mtfranchetto) [inline infrastructure/parallelcluster-ui.yaml:6]: todo: considering we only support the latest PCUI version, this parameter should be removed entirely. Basically, only the latest version of PCUI will be installed by customers. Development deploys still work by using the development parameters (lambda and infrastructure buckets)

File: infrastructure/parallelcluster-ui.yaml [hunk 2]
```diff
@@ -175,13 +181,15 @@ Resources:
...
-          - ${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/${Repository}:${Version}
-          - Repository: !Ref PrivateEcrRepository
-            Version: !Join
-              - '-'
-              - [!Select [2, !Split ['/', !Ref EcrImage]], !Select [3, !Split ['/', !Ref EcrImage]]]
+        S3Bucket: !If
+          - HasDefaultLambda
+          - parallelcluster-ui-releases
+          - !Ref LambdaBucketName
+        S3Key: !Sub '${PCUIVersion}/pcui-lambda.zip'
```
Comment (mtfranchetto) [inline infrastructure/parallelcluster-ui.yaml:192]: todo: same as above comment, considering we only support the latest PCUI version there's no need  to keep multiple artifacts versions in our bucket

File: infrastructure/parallelcluster-ui.yaml [hunk 3]
```diff
@@ -175,13 +181,15 @@ Resources:
...
       Code:
-        ImageUri: !Sub
-          - ${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/${Repository}:${Version}
-          - Repository: !Ref PrivateEcrRepository
-            Version: !Join
-              - '-'
-              - [!Select [2, !Split ['/', !Ref EcrImage]], !Select [3, !Split ['/', !Ref EcrImage]]]
+        S3Bucket: !If
+          - HasDefaultLambda
+          - parallelcluster-ui-releases
```
Comment (mtfranchetto) [inline infrastructure/parallelcluster-ui.yaml:190]: issue: why a new bucket for this zip file? Reuse the same bucket now hosting the YAML files

File: infrastructure/parallelcluster-ui.yaml [hunk 4]
```diff
@@ -35,6 +35,11 @@ Parameters:
     Description: (Optional) S3 bucket where CloudFormation files are stored. Change this parameter only when testing changes made to the infrastructure itself.
     Type: String
     Default: ''
+  LambdaBucketName:
```
Comment (mtfranchetto) [inline infrastructure/parallelcluster-ui.yaml:38]: todo: this new parameter is not needed. For development deployments, the zip file can be published to the same `InfrastructureBucket` where YAML files are present and loaded from there. I suggest to get rid of it and just use `InfrastructureBucket + pcui-lambda.zip` to build the path

**Review by mtfranchetto — CHANGES_REQUESTED** [review]

File: scripts/build_lambda_archive.sh [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+cp -r "${PROJECT_ROOT}/api" ./api 1>&2
+# remove tests from backend code
+rm -r ./api/tests 1>&2
+
+
+# Add main webapp file
+cp "${PROJECT_ROOT}/app.py" ./app.py
+
+print_err "Generating the archive '$ARCHIVE_NAME'"
+zip -r "$ARCHIVE_NAME" . 1>&2
```
Comment (mtfranchetto) [inline scripts/build_lambda_archive.sh:89]: question: is the `zip` command available on Actions?

Comment (BarcoMasile) [reply scripts/build_lambda_archive.sh]: It seems to be installed using the [action-zip](https://github.com/marketplace/actions/action-zip), so we should be good to go

**Review by mtfranchetto — APPROVED** [review]

**Pull request merged**